README.md
# NodeBB Widget: Stats

Overall stats about life on the forum. Some metrics are number of posts, users, topics. Also several additional user related features like - currently online users and who have visited today.

![Version](https://img.shields.io/npm/v/nodebb-widget-ns-stats.svg)
![Dependencies](https://img.shields.io/bithound/dependencies/github/nicolassiver/nodebb-widget-ns-stats.svg)
[![Maintainability](https://api.codeclimate.com/v1/badges/7b2349bd4833c108799c/maintainability)](https://codeclimate.com/github/NicolasSiver/nodebb-widget-ns-stats/maintainability)

<!-- START doctoc generated TOC please keep comment here to allow auto update -->
<!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->
 

- [Preview](#preview)
- [Implementation](#implementation)

<!-- END doctoc generated TOC please keep comment here to allow auto update -->

## Preview

![Widget Preview](screenshot.png)

## Implementation

- Users are sorted by `username` in online list
- Users are sorted by time of appearance in today list
- Online list response is cached for 60 seconds
- Online list is limited to 128 records
- Online status is true, if user was around for last 15 minutes
- General stats are cached for 1 day


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [3.0.0] - 2019-12-16

- Changed compatibility with NodeBB 1.11.x

## [2.0.0] - 2018-04-11

- Changed compatibility with NodeBB 1.8.x
- Changed dependencies to use latest
- Changed widget hook to use HTML output

## [1.0.3] - 2016-04-02

- Compatibility with v1 NodeBB

## [1.0.2] - 2016-02-15

- Fix #2: handle null objects from NodeBB's User module

## [1.0.1] - 2016-01-18

- Possible fix with redis database and visitors list

## [1.0.0] - 2016-01-06

- initial release

widget/job.js
(function (Job) {
    'use strict';

    var async = require('async');

    var logger = require('./logger'),
        nodebb = require('./nodebb'),
        users  = require('./users'),
        db     = nodebb.db;

    var working     = false,
        startAt     = null,
        endAt       = null,

        // Memory storage
        usersCount  = 0,
        topicsCount = 0,
        postsCount  = 0;

    Job.getCurrentStats = function (done) {
        return done(null, {
            users : usersCount,
            topics: topicsCount,
            posts : postsCount
        });
    };

    Job.start = function (done) {
        if (working) {
            return done(new Error('Job is in progress. You can not start another'));
        }

        working = true;
        startAt = Date.now();
        endAt = null;

        async.parallel({
            users : async.apply(db.getObjectField, 'global', 'userCount'),
            topics: async.apply(db.getObjectField, 'global', 'topicCount'),
            posts : async.apply(db.getObjectField, 'global', 'postCount')
        }, function (error, result) {
            working = false;
            endAt = Date.now();

            if (error) {
                return done(error);
            }

            if (usersCount != 0) {
                users.reset(function (error) {
                    if (error) {
                        return logger.log('error', 'Can not reset users. %s', error);
                    }
                });
            }

            usersCount = result.users;
            topicsCount = result.topics;
            postsCount = result.posts;

            done(null);
        });
    };

})(module.exports);


widget/index.js
(function (Widget) {
    'use strict';

    var async = require('async');

    var controller = require('./controller'),
        logger     = require('./logger');

    Widget.hooks = {
        actions: {
            pluginDeactivate: function (pluginId) {
                if (pluginId === require('../plugin.json').id) {
                    controller.dispose(function (error) {
                        if (error) {
                            return logger.log('error', '%s', error);
                        }
                        logger.log('verbose', 'Plugin deactivation: everything is disposed');
                    });
                }
            },

            /**
             * @param payload will include 2 fields, 'uid' and 'timestamp'
             */
            userOnline: function (payload) {
                controller.registerOnlineUser(payload.uid, function (error) {
                    if (error) {
                        return logger.log('error', '%s', error);
                    }
                });
            }
        },
        filters: {
            clearRequireCache: function (data, callback) {
                controller.dispose(function (error) {
                    if (error) {
                        logger.log('error', '%s', error);
                        return callback(error);
                    }
                    callback(null, data);
                });
            },

            getWidgets: controller.getWidgets,

            renderWidget: controller.renderWidget
        },
        statics: {
            load: function (params, callback) {
                async.series([
                    async.apply(controller.setParams, params),
                    async.apply(controller.loadTemplates),
                    async.apply(controller.loadData),
                    async.apply(controller.deferStart),
                    async.apply(controller.setupCron)
                ], function (error) {
                    if (error) {
                        return callback(error);
                    }
                    logger.log('verbose', 'Widget is initiated successfully');
                    callback(null);
                });
            }
        }
    };

})(module.exports);


widget/nodebb.js
(function (Module, NodeBB) {
    'use strict';

    Module.exports = {
        adminSockets : NodeBB.require('./src/socket.io/admin').plugins,
        cache        : NodeBB.require('./src/posts/cache'),
        db           : NodeBB.require('./src/database'),
        groups       : NodeBB.require('./src/groups'),
        meta         : NodeBB.require('./src/meta'),
        pluginSockets: NodeBB.require('./src/socket.io/plugins'),
        postTools    : NodeBB.require('./src/posts/tools'),
        serverSockets: NodeBB.require('./src/socket.io').server.sockets,
        settings     : NodeBB.require('./src/settings'),
        socketIndex  : NodeBB.require('./src/socket.io/index'),
        topics       : NodeBB.require('./src/topics'),
        user         : NodeBB.require('./src/user'),

        utils  : NodeBB.require('./public/src/utils'),
        helpers: NodeBB.require('./src/controllers/helpers'),

        /**
         * List is incomplete
         *
         * base_dir: '/path/to/NodeBB',
         * themes_path: '/path/to/NodeBB/node_modules',
         * views_dir: '/path/to/NodeBB/public/templates',
         * version: 'NodeBB Version',
         * url: 'http://localhost:4567',
         * core_templates_path: '/path/to/NodeBB/src/views',
         * base_templates_path: '/path/to/NodeBB/node_modules/nodebb-theme-vanilla/templates',
         * upload_path: '/public/uploads',
         * relative_path: '',
         * port: '4567',
         * upload_url: '/uploads/',
         * theme_templates_path: '/path/to/NodeBB/node_modules/nodebb-theme-lavender/templates',
         * theme_config: '/path/to/NodeBB/node_modules/nodebb-theme-lavender/theme.json',
         * NODE_ENV: 'development'
         */
        nconf    : NodeBB.require('nconf'),
        passport : NodeBB.require('passport'),
        express  : NodeBB.require('express'),
        templates: NodeBB.require('benchpressjs')
    };
})(module, require.main);


widget/controller.js
(function (Controller) {

    var async   = require('async'),
        CronJob = require('cron').CronJob,
        fs      = require('fs'),
        path    = require('path');

    var job         = require('./job'),
        logger      = require('./logger'),
        nodebb      = require('./nodebb'),
        nconf       = nodebb.nconf,
        templatesJs = nodebb.templates,
        Templates   = require('./templates'),
        users       = require('./users');

    var app       = null,
        cronJob   = null,
        templates = null;

    Controller.deferStart = function (done) {
        // Postpone initial Job,
        setTimeout(function defer() {
            job.start(function (error) {
                if (error) {
                    return logger.log('error', 'Initial error has occurred. %s', error);
                }
                logger.log('verbose', 'Initial job is started');
            });
        }, 500 + Math.random() * 2500);
        done(null);
    };

    Controller.dispose = function (done) {
        if (cronJob !== null) {
            logger.log('warn', 'Cron Job is disposed');
            cronJob.stop();
            cronJob = null;
        }
        done(null);
    };

    Controller.getWidgets = function (widgets, done) {
        widgets.push({
            name       : 'Stats',
            widget     : 'ns_stats',
            description: "Different general metrics about your forum: posts, topics, users",
            content    : templates[Templates.SETTINGS].data
        });

        done(null, widgets);
    };

    Controller.loadData = function (done) {
        users.restore(done);
    };

    Controller.loadTemplates = function (done) {
        if (templates !== null) {
            logger.log('warn', 'Templates are already loaded');
            return done(null);
        }

        templates = {};
        templates[Templates.SETTINGS] = {uri: 'widgets/stats/settings.tpl', data: undefined};
        templates[Templates.VIEW] = {uri: 'widgets/stats/view.tpl', data: undefined};

        async.each(Object.keys(templates), function (name, next) {
            var template = templates[name];
            fs.readFile(path.resolve(__dirname, '../public/templates', template.uri), function (error, content) {
                if (error) {
                    logger.log('error', 'Template Error has occurred, message: %s', error.message);
                    return next(error);
                }
                template.data = content.toString();
                logger.log('verbose', 'Widget Template "%s" is loaded', name);
                next(null);
            });
        }, done);
    };

    Controller.registerOnlineUser = function (uid, done) {
        users.online(parseInt(uid), done);
    };

    Controller.renderWidget = function (widget, done) {
        // Settings to access
        // widget.data.something;

        async.parallel({
            stats : async.apply(job.getCurrentStats),
            today : async.apply(users.getToday),
            online: async.apply(users.getOnline)
        }, function (error, data) {
            if (error) {
                return done(error);
            }

            data.relative_path = nconf.get('relative_path');
            data.visitorsTitle = widget.data.visitorsTitle || 'Visitors Today';
            data.onlineTitle = widget.data.onlineTitle || 'Who\'s Online';

            templatesJs.compileParse(templates[Templates.VIEW].data, data, function (error, output) {
                widget.html = output;
                done(error, widget);
            });
        });
    };

    Controller.setParams = function (params, done) {
        var router      = params.router,
            middleware  = params.middleware,
            controllers = params.controllers;

        app = params.app;
        done(null);
    };

    Controller.setupCron = function (done) {
        if (cronJob !== null) {
            logger.log('warn', 'Cron Job is installed already');
            return done(null);
        }

        // Runs every day
        // at 00:15 AM
        cronJob = new CronJob('00 15 0 * * *', function () {
                logger.log('verbose', 'Job is launched');
                job.start(function (error) {
                    if (error) {
                        return logger.log('error', '%s', error);
                    }
                    logger.log('verbose', 'Job is finished');
                });
            }, null, true
        );

        done(null);
    };

})(module.exports);


widget/users.js
(function (Users) {
    'use strict';

    var async = require('async');

    var logger = require('./logger'),
        nodebb = require('./nodebb'),
        db     = nodebb.db,
        user   = nodebb.user;

    var online        = [],
        onlineIndex   = [],
        onlineCache   = 60000,  // 1 minute
        onlineStatus  = 900000, // 15 minutes
        onlineLimit   = 128,
        lastRequestAt = 0,
        storageKey    = 'widget:ns:stats',
        today         = [],
        todayIndex    = [];

    /**
     * Reduce cost of the operation via 60 sec cache
     * @param done {function}
     */
    Users.getOnline = function (done) {
        var now = Date.now();
        if (now - lastRequestAt > onlineCache) {
            var usersOnline, usersToFetch, usersInCache;
            async.waterfall([
                async.apply(db.getSortedSetRevRangeByScore, 'users:online', 0, onlineLimit, now, now - onlineStatus),
                function findNewIds(uids, next) {
                    usersOnline = uids.map(function (uid) {
                        return parseInt(uid);
                    });
                    usersInCache = [];
                    usersToFetch = usersOnline.filter(function (uid) {
                        var cache = onlineIndex.indexOf(uid) != -1;
                        if (cache) {
                            usersInCache.push(uid);
                        }
                        return !cache;
                    });
                    next(null, usersToFetch);
                },
                function fetchIfNeeded(uids, next) {
                    user.getUsersData(uids, next)
                },
                function composeUserList(users, next) {
                    lastRequestAt = now;
                    online = users.concat(online.filter(function (user) {
                        return usersInCache.indexOf(parseInt(user.uid)) != -1;
                    }));
                    online.sort(function (userA, userB) {
                        return userA.username.localeCompare(userB.username);
                    });
                    onlineIndex = usersOnline;

                    next(null, online);
                }
            ], done);
        } else {
            return done(null, online);
        }
    };

    Users.getToday = function (done) {
        return done(null, today);
    };

    Users.online = function (uid, done) {
        if (todayIndex.indexOf(uid) == -1) {
            user.getUserData(uid, function (error, userData) {
                if (error) {
                    return done(error);
                }
                todayIndex.push(uid);
                today.push(userData);
                logger.log('verbose', 'User %d is online', uid);

                db.setObject(storageKey, {today: todayIndex.join(',')}, done);
            });
        } else {
            done(null);
        }
    };

    Users.reset = function (done) {
        today.length = 0;
        todayIndex.length = 0;
        db.delete(storageKey, done);
    };

    Users.restore = function (done) {
        async.waterfall([
            async.apply(db.getObject, storageKey),
            function fetchIfNeeded(data, next) {
                if (data != null && 'today' in data) {
                    next(null, data.today.split(',').map(function (uid) {
                        return parseInt(uid);
                    }));
                } else {
                    next(null, []);
                }
            },
            function fetchUsers(uids, next) {
                user.getUsersData(uids, next);
            },
            function compose(users, next) {
                users = users.filter(function (user) {
                    return !!user;
                });
                today = users;
                todayIndex = users.map(function (user) {
                    return parseInt(user.uid);
                });
                next(null);
            }
        ], done);
    };

})(module.exports);
